compiler/rustc_mir/src/const_eval/fn_queries.rs

```diff
@@ -48,9 +48,12 @@ fn is_const_fn_raw(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
     if let hir::Node::ForeignItem(hir::ForeignItem { kind: hir::ForeignItemKind::Fn(..), .. }) =
         node
     {
-        // Intrinsics use `rustc_const_{un,}stable` attributes to indicate constness. All other
-        // foreign items cannot be evaluated at compile-time.
-        if let Abi::RustIntrinsic | Abi::PlatformIntrinsic = tcx.hir().get_foreign_abi(hir_id) {
+        // Intrinsics use `rustc_const_{un,}stable` attributes to indicate constness.
+        // `panic_impl` also does in order to pass const checks (it is never evaluated at compile
+        // time).
+        if let Abi::Rust | Abi::RustIntrinsic | Abi::PlatformIntrinsic =
```

Comment:
`Abi::Rust` is not an intrinsic ABI, this looks wrong.

Reply:
Ah I assume this is for the `panic_impl`?

This is a rather bad hack, I'd strongly prefer it we can avoid it. I have no idea if it has adverse side-effects.

Reply:
I tried to special-case the `panic_impl` lang item only, but it is `None` in libcore, despite the `#[lang = "panic_impl"]` on `panic_impl`.

Reply:
Hm... but now that you are allowing a stable ABI here (as in, an ABI that can be used by stable code), the ABI check just does not make any sense any more this way; we might as well remove it entirely and always call `lookup_const_stability`.

And then we need to somehow be sure that this cannot be exploited by user code, and that is the part I am unsure about.